File: 289-game-of-life/game-of-life.js
/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var gameOfLife = function(board) {
    let directions = [[-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1], [-1, -1]]
    let rows = board.length
    let columns = board[0].length

    function isValid(i, j) {
        if (i < 0 || j < 0 || i == rows || j == columns) return false
        if (board[i][j] < 1) return false
        return true
    }

    function countNeigbors(i, j) {
        let count = 0
        for (let [x, y] of directions) {
            new_x = x + i
            new_y = j + y
            if (isValid(new_x, new_y)) {
                count += 1
            }
        }
        return count
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            let cell = board[i][j]
            let count = countNeigbors(i, j)
            if (cell == 1) {
                if (!(count == 2 || count == 3)) {
                    board[i][j] = 2
                }

            }
            else if (cell == 0) {
                if (count == 3) {
                    board[i][j] = -1
                }
            }
        }
    }

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            let cell = board[i][j]
            if (cell == 2) board[i][j] = 0
            else if (cell == -1) board[i][j] = 1 
        }
    }

};

File: 289-game-of-life/game-of-life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        # iterate through the grid
        # count the cell of the neighbors
        # if cell is live
        # if count of neigbors is not equal to two or 3, the cell dies (s)
        # if the cell is dead
        # if the count of neighbors is equal to 3, the cell lives on
        rows = len(board)
        columns = len(board[0])
        directions = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1)]


        def isValid(i, j):
            if (i < 0 or j < 0 or i == rows or j == columns):
                return False
            return True

        def countNeighbors(i, j):
            count = 0
            
            for x, y in directions:
                new_x = x + i
                new_y = y + j

                if (isValid(new_x, new_y)):
                    if (board[new_x][new_y] >= 1):
                        count += 1

            return count

        for i in range(rows):
            for j in range(columns):
                cell = board[i][j]

                count = countNeighbors(i, j)
                if (cell == 1):
                    if (count not in [2, 3]):
                        board[i][j] = 2
                elif (cell == 0):
                    if (count == 3):
                        board[i][j] = -1

        for i in range(rows):
            for j in range(columns):
                cell = board[i][j]

                if (cell == -1):
                    board[i][j] = 1
                elif (cell == 2):
                    board[i][j] = 0
                        